File: trunk/vs2010/PlateScheduler/PlateScheduler.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using BioNex.Shared.DeviceInterfaces;
using BioNex.Shared.IError;
using BioNex.Shared.Location;
using BioNex.Shared.PlateWork;
using BioNex.Shared.TaskListXMLParser;
using BioNex.SynapsisPrototype;
using log4net;
using System.Text;
using BioNex.Shared.LibraryInterfaces;

namespace BioNex.PlateScheduler
{
    [ PartCreationPolicy( CreationPolicy.Shared)]
    [ Export( typeof( IPlateScheduler))]
    [Export(typeof(IReportsStatus))]
    public class PlateScheduler : IPlateScheduler
    {
        // ----------------------------------------------------------------------
        // properties.
        // ----------------------------------------------------------------------
        public IRobotScheduler RobotScheduler { get; protected set; }
        protected DeviceManager DeviceManager { get; set; }
        protected IError ErrorInterface { get; set; }
        protected ConcurrentQueue< Worklist> WorklistQueue { get; set; }
        protected Thread PlateSchedulerThread { get; set; }
        protected ManualResetEvent PlateSchedulerThreadStopEvent { get; set; }

        // map of destination plate reference to worklist name
        public Dictionary<ActivePlate,string> DestinationWorklistMap { get; private set; }

        // ----------------------------------------------------------------------
        // members.
        // ----------------------------------------------------------------------
        protected static readonly ILog Log = LogManager.GetLogger( typeof( PlateScheduler));

        // ----------------------------------------------------------------------
        // constructors.
        // ----------------------------------------------------------------------
        [ ImportingConstructor]
        public PlateScheduler( [ Import] IRobotScheduler robot_scheduler, [ Import] DeviceManager device_manager, [ Import] IError error_interface)
        {
            RobotScheduler = robot_scheduler;
            DeviceManager = device_manager;
            ErrorInterface = error_interface;
            WorklistQueue = new ConcurrentQueue< Worklist>();
            PlateSchedulerThreadStopEvent = new ManualResetEvent( false);

            DestinationWorklistMap = new Dictionary< ActivePlate, string>();
        }

        // ----------------------------------------------------------------------
        // methods.
        // ----------------------------------------------------------------------
        public void StartScheduler()
        {
            PlateSchedulerThread = new Thread( PlateSchedulerThreadRunner){ Name = GetType().ToString(), IsBackground = true};
            PlateSchedulerThread.Start();
        }
        // ----------------------------------------------------------------------
        public void StopScheduler()
        {
            PlateSchedulerThreadStopEvent.Set();
            if( PlateSchedulerThread != null){
                PlateSchedulerThread.Join();
            }
        }
        // ----------------------------------------------------------------------
        public void EnqueueWorklist( Worklist worklist)
        {
            WorklistQueue.Enqueue( worklist);
            IEnumerable< DeviceInterface> bb_devices = from device in DeviceManager.DevicePluginsAvailable.Values
                                                       where device.ProductName == "Bumblebee"
                                                       select device;
            Debug.Assert( bb_devices.Count() == 1);
            PlateSchedulerDeviceInterface bb = bb_devices.FirstOrDefault() as PlateSchedulerDeviceInterface;
            Debug.Assert( bb != null);
            bb.EnqueueWorklist( worklist);
        }
        // ----------------------------------------------------------------------
        protected void PlateSchedulerThreadRunner()
        {
            while( !PlateSchedulerThreadStopEvent.WaitOne( 0)){
                Thread.Sleep( 100);
                Worklist worklist;
                bool worklist_dequeued = WorklistQueue.TryDequeue( out worklist);
                if( worklist != null){
                    DoWorklist( worklist);
                }
            }
        }
        // ----------------------------------------------------------------------
        protected void DoWorklist( Worklist worklist)
        {
            IList< ActivePlateFactory> active_plate_factories = new List< ActivePlateFactory>();
            active_plate_factories.Add( new ActiveSourcePlateFactory( worklist));
            active_plate_factories.Add( new ActiveDestinationPlateFactory( worklist));
            // active_plate_factories.Add( new ActiveTipboxFactory( worklist));

            while(( active_plate_factories.Sum( apf => apf.NumberOfPlatesToCreate) > 0) ||
                  ( ActivePlate.ActivePlates.Count > 0))
            {
                Thread.Sleep( 50);

                // DKM 2012-01-28 this is the region where ActivePlate.ActivePlates gets modified, so protect with
                //                a lock so GetStatus() is safe.
                lock( ActivePlate.ActivePlates) {
                    // remove finished active plates:
                    IList< ActivePlate> finished_plates = ActivePlate.ActivePlates.Where( ap => ap.IsFinished()).ToList();
                    foreach( ActivePlate finished_plate in finished_plates){
                        Log.InfoFormat( "Removed {0}", finished_plate);
                    }
                    ActivePlate.ActivePlates = ActivePlate.ActivePlates.Except( finished_plates).ToList();

                    // release active plates as necessary:
                    foreach( ActivePlateFactory active_plate_factory in active_plate_factories){
                        ActivePlate active_plate = active_plate_factory.TryReleaseActivePlate();
                        if( active_plate != null){
                            ActivePlate.ActivePlates.Add( active_plate);
                            // DKM 2011-11-03 refs #544: now associate a destination plate with a worklist so that we can
                            //                display the worklist(s) associated with a particular cart.  This doesn't
                            //                work because the plate barcode isn't known at this point!!!
                            if( active_plate is ActiveDestinationPlate) {
                                lock( DestinationWorklistMap) {
                                    DestinationWorklistMap[active_plate] = worklist.Name;
                                }
                            }
                        }
                    }
                }

                // advance active plates as necessary:
                foreach( ActivePlate active_plate in ActivePlate.ActivePlates){
                    if( ActivePlate.ActivePlates.Count( plate => plate != active_plate && plate.GetCurrentToDo() == active_plate.GetCurrentToDo() && plate.InstanceIndex < active_plate.InstanceIndex) > 0){
                        continue;
                    }
                    if( active_plate.Busy){
                        continue;
                    }
                    PlateTask current_to_do = active_plate.GetCurrentToDo();
                    if( current_to_do == null){
                        continue;
                    }
                    // insert criteria on whether or not to do task here.
                    // TRUE.
                    // find an available location for the task.
                    string device_name = current_to_do.DeviceType;
                    // START INSERTING REAL DEVICE MANAGER INTO PLAY HERE.

                    IDictionary< string, DeviceInterface> devices_of_type = ( from kvp in DeviceManager.DevicePluginsAvailable
                                                                              where kvp.Value.ProductName == device_name
                                                                              select kvp).ToDictionary( x => x.Key, x => x.Value);
                    if( devices_of_type.Count == 0){
                        throw new Exception( "The device type needed is not available in the device manager.");
                    }
                    bool device_scheduled = false;
                    bool stacker_occupied = false;
                    foreach( KeyValuePair< string, DeviceInterface> kvp in devices_of_type){
                        PlateSchedulerDeviceInterface device = kvp.Value as PlateSchedulerDeviceInterface;
                        if( device == null){
                            throw new Exception( "The device type needed is not plate scheduler compliant.");
                        }
                        if(( device is StackerInterface) && ( device as AccessibleDeviceInterface).PlateLocationInfo.First().Occupied.WaitOne( 0)){
                            stacker_occupied = true;
                        }
                        PlateLocation location = device.GetAvailableLocation( active_plate);
                        if( location == null){
                            continue;
                        }
                        if( !device.ReserveLocation( location, active_plate)){
                            continue;
                        }
                        Log.DebugFormat( "Available location {0}.{1} reserved for {2}", device, location, active_plate);

                        // commit active plate to a job:
                        active_plate.PlateIsFree.Reset();
                        if( active_plate.CurrentLocation == null){
                            // we're sourcing the plate for the first time.
                            // no robot move needed.
                            Log.DebugFormat( "Sourcing plate {0} at device {1}", active_plate, device);
                            active_plate.CurrentLocation = location;
                            active_plate.DestinationLocation = location;
                        } else{
                            // we're moving the plate from one location to another to perform the task at hand.
                            Log.DebugFormat( "Queueing up robot job to move plate {0} to device {1}", active_plate, device);
                            active_plate.DestinationLocation = location;
                            Debug.Assert( !active_plate.DestinationLocation.Occupied.WaitOne( 0));
                            RobotScheduler.AddJob( active_plate);
                        }
                        Log.DebugFormat( "Queueing up device job to execute {0} on plate {1}", active_plate.GetCurrentToDo().Command, active_plate);
                        device.AddJob( active_plate);
                        device_scheduled = true;
                        break;
                    }

                    // DKM 2011-10-13 we have a potential issue if the racks aren't loaded.  I was planning on calling LoadStack()
                    //                before each call to Upstack() and Downstack(), but the problem is that these methods will
                    //                never get called if the rack isn't loaded!  So we need to warn the user if we fall through
                    //                the previous loop and don't schedule a job.
                    if( !device_scheduled && !stacker_occupied && devices_of_type.First().Value is StackerInterface){
                        ErrorData error_data = new ErrorData( "No destination plates available in stackers -- please ensure that they are loaded properly", new List< string>{ "Retry"});
                        List< ManualResetEvent> events = new List< ManualResetEvent>();
                        events.AddRange( error_data.EventArray);
                        ErrorInterface.AddError( error_data);
                        int event_index = WaitHandle.WaitAny( events.ToArray());
                        if( error_data.TriggeredEvent == "Retry"){
                        } else{
                        }
                    }
                }
            }
            Log.InfoFormat( "Worklist completed");
            worklist.OnWorklistComplete();
        }

        /// <summary>
        /// The PlateScheduler owns the RobotScheduler, so GetStatus needs to call RobotScheduler's GetStatus method
        /// </summary>
        /// <returns></returns>
        public string GetStatus()
        {
            StringBuilder sb = new StringBuilder();

            // robot scheduler status
            sb.AppendLine( RobotScheduler.GetStatus());

            // plate scheduler status includes WorklistQueue and ActivePlates
            sb.AppendLine( "\tActive plates:");
            lock( ActivePlate.ActivePlates) {
                foreach( ActivePlate plate in ActivePlate.ActivePlates) {
                    sb.AppendLine( plate.GetStatus());
                }
            }

            return sb.ToString();
        }
    }
}

File: scheduler_framework/scheduler/plate_scheduler.py
"""
PlateScheduler - Main orchestrator for laboratory automation scheduling.

This is the core scheduler that:
- Manages worklists
- Creates and tracks ActivePlates
- Schedules tasks on available devices
- Coordinates with RobotScheduler for movements
"""

import threading
import time
import logging
from queue import Queue
from typing import List, Dict, Optional
from .worklist import Worklist
from .tasks import WaitTask
from .active_plate import ActivePlate, ActiveSourcePlate, ActiveDestinationPlate, PlateLocation
from .device_manager import DeviceManager, DeviceInterface, PlateSchedulerDeviceInterface
from .robot_scheduler import RobotScheduler
from .active_plate_factory import ActivePlateFactory, ActiveSourcePlateFactory, ActiveDestinationPlateFactory

logger = logging.getLogger(__name__)


class PlateScheduler:
    """
    Main scheduler for laboratory automation.
    
    Based on the C# PlateScheduler, this orchestrates the entire
    workflow of processing plates through various devices.
    """
    
    def __init__(self, robot_scheduler: RobotScheduler, 
                 device_manager: DeviceManager):
        self.robot_scheduler = robot_scheduler
        self.device_manager = device_manager
        self.worklist_queue: Queue[Worklist] = Queue()
        self.scheduler_thread: Optional[threading.Thread] = None
        self.stop_event = threading.Event()
        self.destination_worklist_map: Dict[ActivePlate, str] = {}
        self._lock = threading.Lock()
    
    def start_scheduler(self):
        """Start the plate scheduler thread"""
        if self.scheduler_thread and self.scheduler_thread.is_alive():
            return
        
        self.stop_event.clear()
        self.scheduler_thread = threading.Thread(
            target=self._scheduler_thread_runner,
            name="PlateScheduler",
            daemon=True
        )
        self.scheduler_thread.start()
        logger.info("PlateScheduler started")
    
    def stop_scheduler(self):
        """Stop the plate scheduler thread"""
        self.stop_event.set()
        if self.scheduler_thread and self.scheduler_thread.is_alive():
            self.scheduler_thread.join(timeout=1.0)
            if self.scheduler_thread.is_alive():
                logger.warning("PlateScheduler thread did not stop within timeout")
        logger.info("PlateScheduler stopped")
    
    def enqueue_worklist(self, worklist: Worklist):
        """Add a worklist to the processing queue"""
        self.worklist_queue.put(worklist)
        logger.info(f"Enqueued worklist: {worklist.name}")
    
    def _scheduler_thread_runner(self):
        """Main scheduler loop - processes worklists"""
        while not self.stop_event.is_set():
            try:
                # Try to get a worklist
                try:
                    worklist = self.worklist_queue.get(timeout=0.1)
                except:
                    continue
                
                # Process the worklist
                self._do_worklist(worklist)
                
            except Exception as e:
                logger.error(f"Error in plate scheduler: {e}", exc_info=True)
                time.sleep(0.1)
    
    def _do_worklist(self, worklist: Worklist):
        """
        Process a worklist.
        
        Creates ActivePlates from the worklist and schedules
        their tasks on available devices.
        """
        logger.info(f"Processing worklist: {worklist.name}")
        
        # Create factories for active plates
        active_plate_factories: List[ActivePlateFactory] = [
            ActiveSourcePlateFactory(worklist),
            ActiveDestinationPlateFactory(worklist)
        ]
        
        # Main processing loop
        while True:
            # Check if we're done
            plates_to_create = sum(factory.number_of_plates_to_create 
                                 for factory in active_plate_factories)
            active_plates_count = len(ActivePlate._active_plates)
            
            if plates_to_create == 0 and active_plates_count == 0:
                break
            
            time.sleep(0.05)  # Small delay to prevent CPU spinning
            
            # Remove finished plates
            with self._lock:
                finished_plates = [ap for ap in ActivePlate._active_plates 
                                 if ap.is_finished()]
                for plate in finished_plates:
                    logger.info(f"Removed finished plate: {plate}")
                    ActivePlate._active_plates.remove(plate)
            
            # Release new active plates from factories
            with self._lock:
                for factory in active_plate_factories:
                    active_plate = factory.try_release_active_plate()
                    if active_plate:
                        ActivePlate._active_plates.append(active_plate)
                        logger.debug(f"Released new active plate: {active_plate}")
                        
                        # Track destination plates
                        if isinstance(active_plate, ActiveDestinationPlate):
                            self.destination_worklist_map[active_plate] = worklist.name
            
            # Advance active plates
            with self._lock:
                active_plates = list(ActivePlate._active_plates)
            
            for active_plate in active_plates:
                # Skip if there's a plate with same task and lower instance index
                same_task_plates = [ap for ap in active_plates 
                                  if ap != active_plate and 
                                  ap.get_current_todo() == active_plate.get_current_todo() and
                                  ap.instance_index < active_plate.instance_index]
                if same_task_plates:
                    continue
                
                # Skip if busy
                if active_plate.busy:
                    continue
                
                # Get current task
                current_task = active_plate.get_current_todo()
                if not current_task:
                    continue
                
                # Handle WaitTask specially (no device needed)
                if isinstance(current_task, WaitTask):
                    if not active_plate.busy:
                        # Start wait task
                        self._handle_wait_task(active_plate, current_task)
                    continue
                
                # Find available device for this task
                device_scheduled = self._schedule_task(active_plate, current_task)
                
                if not device_scheduled:
                    logger.debug(f"Could not schedule task {current_task} for {active_plate}")
        
        logger.info(f"Worklist {worklist.name} completed")
        worklist.on_worklist_complete()
    
    def _handle_wait_task(self, active_plate: ActivePlate, wait_task: WaitTask):
        """
        Handle a wait task by waiting for the specified duration.
        
        This runs in a separate thread to avoid blocking the scheduler.
        """
        if wait_task.completed:
            return
        
        logger.info(f"Starting wait task: {wait_task.duration_seconds}s for {active_plate}")
        active_plate.plate_is_free.clear()  # Mark as busy
        
        def wait_and_complete():
            import time
            time.sleep(wait_task.duration_seconds)
            wait_task.completed = True
            logger.info(f"Wait task completed for {active_plate}")
            active_plate.mark_job_completed()
        
        # Run wait in background thread
        wait_thread = threading.Thread(
            target=wait_and_complete,
            name=f"WaitTask-{active_plate.plate_serial_number}",
            daemon=True
        )
        wait_thread.start()
    
    def _schedule_task(self, active_plate: ActivePlate, 
                      task) -> bool:
        """
        Schedule a task on an available device.
        
        Returns True if task was scheduled, False otherwise.
        """
        device_type = task.device_type
        
        # Get devices of the required type
        devices_of_type = self.device_manager.get_devices_by_type(device_type)
        
        if not devices_of_type:
            logger.error(f"No devices of type {device_type} available")
            return False
        
        # Try each device to find an available location
        for device_name, device in devices_of_type.items():
            if not isinstance(device, PlateSchedulerDeviceInterface):
                logger.error(f"Device {device_name} is not plate scheduler compliant")
                continue
            
            # Get available location
            location = device.get_available_location(active_plate)
            if not location:
                continue
            
            # Try to reserve location
            if not device.reserve_location(location, active_plate):
                continue
            
            logger.debug(f"Reserved location {location.name} on {device_name} for {active_plate}")
            
            # Commit the plate to this job
            active_plate.plate_is_free.clear()
            
            if active_plate.current_location is None:
                # First time sourcing the plate
                logger.debug(f"Sourcing plate {active_plate} at device {device_name}")
                active_plate.current_location = location
                active_plate.destination_location = location
            else:
                # Moving plate to new location
                logger.debug(f"Queueing robot job to move {active_plate} to {device_name}")
                active_plate.destination_location = location
                self.robot_scheduler.add_job(active_plate)
            
            # Queue device job
            logger.debug(f"Queueing device job {task.command} on {active_plate}")
            device.add_job(active_plate)
            
            return True
        
        return False
    
    def get_status(self) -> str:
        """Get status string for monitoring"""
        status = []
        
        # Robot scheduler status
        status.append(self.robot_scheduler.get_status())
        
        # Active plates status
        status.append("\tActive plates:")
        with self._lock:
            for plate in ActivePlate._active_plates:
                status.append(plate.get_status())
        
        return "\n".join(status)

